### PR TITLE
Feature/geostore endpoint

### PR DESCRIPTION
**Pull request by dmannarino**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [x] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

The geostore endpoints are unimplemented
Issue Number: GTC-655

## What is the new behavior?

Geostore endpoints are implemented
-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

There is lots of room for optimization, as I query the database numerous times to pull the data together in Python before finally writing to the DB. However those optimizations can be done later and won't change the in-DB records.

### CODE REVIEW

**Comment by thomas-maschler** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- app/models/orm/user_areas.py  1
- app/models/orm/migrations/versions/12a4acb31f39_.py  1
- app/models/orm/migrations/versions/3593e241c4b8_.py  1
- app/crud/geostore.py  3
- app/models/orm/migrations/versions/4a4df534a8f6_.py  1
         

Clones added
============
- app/models/orm/migrations/versions/12a4acb31f39_.py  8
- app/models/orm/migrations/versions/3593e241c4b8_.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wri/gfw-data-api/pullRequest?prid=5867297&bid=19203533)

File: app/crud/geostore.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    geo: Geostore = Geostore.from_orm(row)
+
+    return hydrate_geostore(geo)
+
+
+async def get_geostore_by_version(dataset, version, geostore_id) -> GeostoreHydrated:
+    sql = db.text(
+        f"""
+        SELECT *
+        FROM ONLY "{dataset}"."{version}"
```
Comment (thomas-maschler) [inline app/crud/geostore.py:42]: I believe `ONLY` means here to exclude descendant tables. We might need to partition some of these tables and hence there will be descendant tables. These tables also have more than the standard geostore columns. I would not use the wildcard `*` but list the fields explicitly. 

To assure that you always use the same fields names across the different queries you can use SQL alchemy core.

```python
columns: List[Column] = [db.column(gfw_geostore_id), db.column.(gfw_geojson), db.column(gfw_bbox), db.column(gfw.area__ha)]

src_table: Table = db.table(version)
src_table.schema = dataset

where_clause: TextClause = db.text("gfw_geostore_id=:geostore_id") 
bind_vals = {"geostore_id": f"{geostore_id}"}
where_clause  = where_clause.bindparams(**bind_vals)

sql: Select = db.select(columns).select_from(src_table).where(where_clause)
```

File: app/routes/geostore/geostore.py [hunk 2]
```diff
@@ -1,47 +1,67 @@
...
-    """Add geostore feature to User geostore."""
-    pass
+async def add_new_geostore(
+    *, request: GeostoreIn, response: ORJSONResponse,
+):
+    """Add geostore feature to user area of geostore."""
+
+    input_data = request.dict(exclude_none=True, by_alias=True)
+
+    new_user_area: GeostoreHydrated = await geostore.create_user_area(**input_data)
```
Comment (thomas-maschler) [inline app/routes/geostore/geostore.py:29]: Need to catch `BadRequestError` and make it an `HTTPException` with status code 400

File: app/routes/geostore/geostore.py [hunk 3]
```diff
@@ -1,47 +1,67 @@
...
     """Retrieve GeoJSON representation for a given geostore ID of a dataset
     version.
 
     Obtain geostore ID from feature attributes.
     """
-    pass
+    result: GeostoreHydrated = await geostore.get_geostore_by_version(
+        dataset, version, geostore_id
+    )
+    return GeostoreResponse(data=result)
```
Comment (thomas-maschler) [inline app/routes/geostore/geostore.py:67]: I wonder if it make sense to move this path operation to routes/dataset/geostore.py?

File: tests/tasks/test_default_assets.py [hunk 4]
```diff
@@ -143,15 +143,15 @@ async def test_vector_source_asset(batch_client, async_client):
...
+                    "field_type": "text",
                     "is_feature_info": False,
                     "is_filter": False,
                 },
                 {
                     "field_name": "gfw_bbox",
                     "field_alias": "gfw_bbox",
                     "field_description": None,
-                    "field_type": "geometry",
+                    "field_type": "ARRAY",
```
Comment (thomas-maschler) [inline tests/tasks/test_default_assets.py:154]: Does the current implementation return the type as Array? I might need to change this later to something like double precision[]. (in a separate PR)

File: tests/routes/test_geostore.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    dataset = "test"
+    source = GEOJSON_NAME
+    version = "v1.1.1"
+    input_data = {
+        "creation_options": {
+            "source_type": "vector",
+            "source_uri": [f"s3://{BUCKET}/{source}"],
+            "source_driver": "GeoJSON",
+            "zipped": False,
```
Comment (thomas-maschler) [inline tests/routes/test_geostore.py:76]: I believe I removed the `zipped` attribute in my latest PR which is already in staging.

File: app/crud/geostore.py (hunk 1 above)
Comment (dmannarino) [reply app/crud/geostore.py]: D'oh. Right you are. Thanks for the SA Core lesson. Changed and pushed.

**Comment by thomas-maschler** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- app/crud/geostore.py  3
- app/models/orm/user_areas.py  1
- app/models/orm/migrations/versions/4a4df534a8f6_.py  1
- app/routes/datasets/geostore.py  1
- app/models/orm/migrations/versions/3593e241c4b8_.py  1
- app/models/orm/migrations/versions/12a4acb31f39_.py  1
         

Clones added
============
- app/crud/geostore.py  2
- app/models/orm/migrations/versions/3593e241c4b8_.py  3
- app/models/orm/migrations/versions/12a4acb31f39_.py  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wri/gfw-data-api/pullRequest?prid=5867297&bid=19203533)

File: app/routes/geostore/geostore.py (hunk 3 above)
Comment (dmannarino) [reply app/routes/geostore/geostore.py]: Yes, I forgot you had re-organized things this way. Done now.

File: app/routes/geostore/geostore.py (hunk 2 above)
Comment (dmannarino) [reply app/routes/geostore/geostore.py]: Done, thank you.

File: tests/routes/test_geostore.py (hunk 5 above)
Comment (dmannarino) [reply tests/routes/test_geostore.py]: Removed.

**Comment by thomas-maschler** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Coverage increased per file
===========================
+ app/routes/geostore/geostore.py  21
         

Complexity increasing per file
==============================
- app/models/orm/user_areas.py  1
- app/routes/datasets/geostore.py  2
- app/models/orm/migrations/versions/4a4df534a8f6_.py  1
- app/models/orm/migrations/versions/3593e241c4b8_.py  1
- app/routes/geostore/geostore.py  1
- app/models/orm/migrations/versions/12a4acb31f39_.py  1
- app/crud/geostore.py  3
         

Clones added
============
- app/models/orm/migrations/versions/86ae41de358d_.py  8
- app/models/orm/migrations/versions/3593e241c4b8_.py  3
- app/models/orm/migrations/versions/12a4acb31f39_.py  8
- app/models/orm/migrations/versions/f46eb3aff594_.py  2
- app/crud/geostore.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wri/gfw-data-api/pullRequest?prid=5867297&bid=19203533)

**Pull request merged**